### ISSUE TITLE
enforce the highest bitrate

### ISSUE THREAD

**Issue by famoser** [truedread/netflix-1080p#19]
Impressive work with the chrome profile fix. I have some other clues which might help you.  

If you want to go for maximum quality, consider to set the bitrate directly, as the high-quality-profile still contains multiple bitrates the cadmium player is allowed to choose from.  

I have automated the enforcing of a certain bitrate. It would be quite easy to use this approach to always enforce the highest bitrate; and this should be stable in the sense that the cadmium player is not touched else than forcing the highest bitrate it provides.

It is however implemented using a rather dirty hack. 

There is a window for choosing the bitrates (probably for netflix devs) accessible with the combination `SHIFT + CONTROL + ALT + 'S'`. One can choose the bitrates, and then can press on "override", effectively enforcing said bitrate(s).

The best approach to automate I've found I to simply replicate this process: First open/close the window (so all necessary variables are populated) and then invoking the "override" routine with an array of bitrates. While I am not quite sure what happens exactly in the override routine, it sure works. 

To implement this, you have to:
- include the code at the bottom into your cadmium playercode
- invoke the key combination described above two times (open/close) after the video started playing. You may enforce with CSS that the user never sees that window
- invoke `get_bitrates()` to get the array of possible bitrates
- invoke `set_bandwith(bitrate)` for the bitrate you want to use (the highest one probably)

If this looks like a good idea to you I'll make a pull request in the following week. 

insert after line 39448:

``` js
window.get_bitrates = function() {
	var res = [];
	for (var a = F.options, c = a.length; c--;) {
		res.push(a[c].value);
	}
	return res;
}
window.set_bandwidth = function(value) {
	D = {};
	D[value] = 1;
	b.li = k;
	l();
	if (a = b.pf) {
		var h = G.value,
			a = a.filter(function (a) {
				return a.id == h;
			})[0],
			c = y.te ? b.wa[n.I.VIDEO].value : b.wa.value;
		a && a != c && (a.Yoa = {
			testreason: "streammanager",
			selreason: "userselection"
		}, y.te ? b.wa[n.I.VIDEO].set(a) : b.wa.set(a));
	}
	f();
	return true;
}
```

**Comment by truedread** [comment]
I don't know if I want to implement this in my extension. All my extension is supposed to be is a simple fix for getting Netflix to play in 1080p. I only want to modify the playercore in ways which achieves this goal. You are more than welcome to implement your feature in a fork for people to use if they choose, though.

**Comment by famoser** [comment]
Whether it makes sense to implement this or not depends on your target: 
- (1) Enable 1080p on chrome, but let the cadmium player decide what to play
- (2) Enforce the highest possible quality

It sounds like right now you are aiming for (1), for this you do not need the bitrate enforcement, therefore I'll close this issue. 
If you want to aim for (2) now or any time in the future (potentially with an optional setting), please reopen, and I'll be glad to make a pull request.

**Comment by famoser** [comment]
Unfortunately not. The settings menu populates quite a lot of variables, and due to the js minification it would have been quite hard to replicate the exact behaviour. 

However, this extension replaces the whole video player javascript to its modified version (which still contains the setting window). Hence, if netflix did not change the way its streams work, or else checks for a modified player javascript, you can simply install this extension to get the setting window back. 
I cannot check this, as I do not own a netflix subscription, but kindly post your results if you find the time to try it.

**Comment by trdesilva** [comment]
Interesting; I installed the extension a few days ago and it's working for me, and I can access the settings menu now. Thanks for your input, though; I wouldn't have the patience to dig around in minified JS to figure this out for myself.

**Comment by ItsNickBarry** [comment]
Do you have any idea how to open the bitrate menu programatically?  I can't find an event listener for the keyboard combination, and simulating a keyboard event doesn't work.

**Comment by famoser** [comment]
Unfortunately not; and as I do not have a netflix account I can not try to find something new at the moment. 

I do not think they have removed this capability though (for their internal quality control), but its probably just available in a less obvious way. 

I'd look for javascript functions they cannot minify, and which process user input in some way. Like clicking 5 time in a row at a specific point on the screen, or writing some magic number to the console.

**Comment by ItsNickBarry** [comment]
With this plugin installed, I meant.  But I figured it out.

The event is attached to `window`, rather than `document`, which was not immediately obvious.

I wrote a Firefox plugin to automatically open the menu and choose the highest bitrate, as long as the menu is available.  The code works on Chrome, too, but I don't work for Google.  The repository is [here](https://github.com/ItsNickBarry/netflix-max-bitrate).
